### DISCUSSION TITLE
[IMPORTANT] Mandatory Procedures to Keep Your EAF Up-To-Date

### DISCUSSION THREAD

**Discussion by MatthewZMD** [emacs-eaf/emacs-application-framework#527] [Show and tell]
In addtion to `git pull` that retrieves the latest changes to the EAF project, there may occasionally be some other procedures needed to run to continue using EAF on your computer, due to the complexity of the project.

This post is dedicated to this purpose, you only need to care if you are *not* a first-time EAF user.

Press **Newest** on ↪️ bottom-right to view the latest post

**Comment by MatthewZMD** [comment]
Users before **January 21st, 2021**:

1. Run `npm install` at the root of the EAF project
2. Run `pip install --user epc`
3. Install `ctable`, `deferred`, and `epc` Emacs packages from MELPA

**Comment by MatthewZMD** [comment]
Users before **January 29th, 2021**:
1. Run `pip install --user retrying`
2. Run `npm install` at the root of the EAF project

**Comment by MatthewZMD** [comment]
Users before **July 31, 2021**:

1. Please change all your instances of `eaf-setq` to `setq`, and change the string values `"true"` to `t`, `"false"` to `nil`, `"5"` to number `5`.

See https://github.com/emacs-eaf/emacs-application-framework/issues/734.
for more information.
    
```
In summary, the EAF’s python applications can now access users’ customized
variables with the get_emacs_var module, eliminating the need to store
them in ‘eaf-var-list’ and thus, ‘eaf-setq’ is no longer necessary.
```

2. However, change the `eaf-*-dark-mode` variables from `t` to `"force"`, they're now `"force"`, `"follow"`, and anything else represents false.

**Comment by MatthewZMD** [comment]
Users before **August 6, 2021**:

1. Linux users, run `./install-eaf.sh --ignore-sys-deps`

**Comment by manateelazycat** [comment]
If you define a JavaScript function "myFunction", you just need binding eaf app keymap with "js_my_function", EAF will let you call JavaScript from Elisp side directly.

**Comment by MatthewZMD** [comment]
**SUPER IMPORTANT**

Users before **August 11, 2021**:

Since its birth in 2018, the EAF project has grown pretty big with more than 20 applications to date, we understand that most people will not use them all, it is unnecessary to ship everything together and force users to install all dependencies.

Therefore, we decide to split all applications into their own repository located in the [emacs-eaf](https://github.com/emacs-eaf) Github Organization, this resulted in the EAF core repo having a new link. Although the old `manateelazycat/emacs-application-framework` will redirect to the new link, it is highly recommended to update the `git` remote URL to avoid confusion:
```
git remote set-url origin https://github.com/emacs-eaf/emacs-application-framework.git                 
```

Additionally, we introduce the all-new `install-eaf.py` installation script, this gives you the flexibility to choose which apps to install, and only install their dependencies. If you want, on the latest master, *delete* your `emacs-application-framework/app` directory and run `install-eaf.py` to choose your interested applications:
```
python ./install-eaf.py
```

On a side note, developing your own EAF apps has never been this easier, it doesn't have to be within the emacs-eaf organization, you can put it in your own repository. Simply follow these steps:
1. Ensure the `dependencies.json` in your application repo is following the standard in other EAF applications.
2. Add your repo's description and Git URL to the EAF core `applications.json`.
3. Submit a PR at EAF core repo:
https://github.com/emacs-eaf/emacs-application-framework

**Comment by manateelazycat** [comment]
Emacs Application Framework has split all applications to https://github.com/emacs-eaf/, you need use below commands install Newest version:

1. Download emacs-application-framework
2. `chmod +x ./install-eaf.py ; ./install-eaf.py`
3. Then based on which apps you install, add below the `require` commands in your .emacs
```
(require 'eaf-js-video-player)
(require 'eaf-org-previewer)
(require 'eaf-image-viewer)
(require 'eaf-music-player)
(require 'eaf-file-sender)
(require 'eaf-file-browser)
(require 'eaf-airshare)
(require 'eaf-system-monitor)
(require 'eaf-browser)
(require 'eaf-jupyter)
(require 'eaf-pdf-viewer)
(require 'eaf-video-player)
(require 'eaf-demo)
(require 'eaf-netease-cloud-music)
(require 'eaf-mindmap)
(require 'eaf-terminal)
(require 'eaf-markdown-previewer)
(require 'eaf-vue-demo)
(require 'eaf-file-manager)
(require 'eaf-camera)
```

**Comment by MatthewZMD** [comment]
Users before **August 16st, 2021**:

If you're using `js-video-player`, please remove it.
    
Upon `git pull`, delete the `emacs-application-framework/js-video-player` directory and remove `(require 'eaf-js-video-player)` from your `init.el`
    
We've decided to drop the support for this application, because `plyr.js` library is not stable and does not work anymore, using the existing `video-player` application is enough.

**Comment by manateelazycat** [comment]
I have push some patches to make EAF to support PyQt6

https://github.com/emacs-eaf/emacs-application-framework/commit/c3ab6a600d2fce562bd15c0e0249604d7974bbac
https://github.com/emacs-eaf/eaf-video-player/commit/e8cb927a5010a24dcb6665ac1dc6621de0d037c8
https://github.com/emacs-eaf/eaf-terminal/commit/5ec18e2bb5379f7fc858a9c83ba3a8155dc3a781
https://github.com/emacs-eaf/eaf-system-monitor/commit/1c22e078eb7a9c590b6cb52ff460921aa23ab107
https://github.com/emacs-eaf/eaf-rss-reader/commit/45d351fed413c5873ebd01e44f552386d0f0fce6
https://github.com/emacs-eaf/eaf-pdf-viewer/commit/49ad2d72ebc291fc7530323f91579eb7a7d887d7
https://github.com/emacs-eaf/eaf-org-previewer/commit/f6d6e21fa035a7072b84bf929338ea70f1310041
https://github.com/emacs-eaf/eaf-netease-cloud-music/commit/0beb7dd17a4e704ab977ebf3c6da29e50af67e69
https://github.com/emacs-eaf/eaf-music-player/commit/34ca2a9b5f6fb49e22581c84d7b5ecef61f4fe87
https://github.com/emacs-eaf/eaf-mindmap/commit/f972ceb56eeccec5754eb36c2340e78fdaf4c240
https://github.com/emacs-eaf/eaf-mermaid/commit/f0ab2dc938c1efbd980d2e3fdfdb12bafc3e4b4c
https://github.com/emacs-eaf/eaf-markdown-previewer/commit/a0bde893eae52148fbf1468e7fcad8c530cf2cad
https://github.com/emacs-eaf/eaf-jupyter/commit/7769d3aa0d4d4b78e879b0ea36615b0564d8b1b0
https://github.com/emacs-eaf/eaf-image-viewer/commit/3356a181917a8847d440556744b38961bd34f201
https://github.com/emacs-eaf/eaf-git/commit/2b335b74a6bed6d8a5f07dd80b0a7ee7ca760c9b
https://github.com/emacs-eaf/eaf-file-sender/commit/6bc29a2194bcb8c8378d063c5019143e2d77cc32
https://github.com/emacs-eaf/eaf-file-manager/commit/4c9bfbe10957f42f4d802890de31bb0331ba5f52
https://github.com/emacs-eaf/eaf-file-browser/commit/e3860cdf6485b76b08f1c2beac7283985a04f4fe
https://github.com/emacs-eaf/eaf-demo/commit/c2be29747e17c06c67b4a116391309e0fec97afc
https://github.com/emacs-eaf/eaf-camera/commit/143a437b58c000e3cb84577b1e458a72d84c0507
https://github.com/emacs-eaf/eaf-browser/commit/0b4e57dfbf7230ea81c0e14066ef0a11950c32e4
https://github.com/emacs-eaf/eaf-airshare/commit/e78b3f2a6dc6bb3de4bf71489725ab5beab373d2

You just need run script ./install-eaf.py to upgrade newest version!

**NOTE: please make sure your operating system support Qt6 and PyQt6 before upgrade!!!**

**Comment by manateelazycat** [comment]
EAF now uses pip to install PyQt6 for all Linux distributions, avoiding some distributions repositories too old that not include PyQt6.

But the PIP version of PyQt6 has a bug that crashes directly when login the website, similar to the following error:

````
Thread 1 "python3" received signal SIGTRAP, Trace/breakpoint trap.
0x00007fffede3edc1 in ?? () from /usr/lib/python3.10/site-packages/PyQt6/Qt6/lib/libQt6WebEngineCore.so.6
#0 0x00007fffede3edc1 in () at /usr/lib/python3.10/site-packages/PyQt6/Qt6/lib/libQt6WebEngineCore.so.6
#1 0x0000555555a9dc98 in ()
#2 0x00000000000000f0 in ()
#3 0x00007fffffffc6a0 in ()
#4 0x000055555616c2d8 in ()
#5 0x00007fffffffc5a0 in ()
#6 0x00007fffedcb3801 in () at /usr/lib/python3.10/site-packages/PyQt6/Qt6/lib/libQt6WebEngineCore.so.6
#7 0x00007fffffffc550 in ()
#8 0x00007fffffffc5f0 in ()
#9 0x00000000000000e0 in ()
#10 0x0ec4e89c28622e00 in ()
#11 0x00007fff442e1750 in ()
#12 0xffffffffffffff80 in ()
#13 0x0000000000000000 in ()
````

The solution is to install the Qt6WebEngine package in the Linux distribution, and then symlink the libQt6WebEngineCore.so.6 file to the PyQt6 directory to fix the crash. Here are the solutions for each Linux distribution:

1. Fedora, install ```qt6-qtwebengine-libs``` and run ```sudo ln -sf /usr/lib64/libQt6WebEngineCore.so.6 /usr/lib/python3.10/site-packages/PyQt6/ Qt6/lib/libQt6WebEngineCore.so.6```
2. Gentoo, install the package including ```libQt6WebEngineCore.so.6```, soft link the file to /usr/lib/python3.10/site-packages/PyQt6/Qt6/lib/libQt6WebEngineCore.so.6
3. Debian/Ubuntu, install ```libqt6webenginecore6``` and run ```sudo ln -sf /usr/lib/x86_64-linux-gnu/libQt6WebEngineCore.so.6 /usr/lib/python3.10/site- packages/PyQt6/Qt6/lib/libQt6WebEngineCore.so.6```
4. Arch, install ```qt6-webengine``` and run ```sudo ln -sf /usr/lib/libQt6WebEngineCore.so.6 /usr/lib/python3.10/site-packages/PyQt6/Qt6/ lib/libQt6WebEngineCore.so.6````
5. OpenBSD, install the package including ```libQt6WebEngineCore.so.6```, soft link the file to /usr/lib/python3.10/site-packages/PyQt6/Qt6/lib/libQt6WebEngineCore.so.6
6. Suse, install ```libQt6WebEngineCore6``` and run ```sudo ln -sf /usr/lib/libQt6WebEngineCore.so.6 /usr/lib/python3.10/site-packages/PyQt6/Qt6/lib/ libQt6WebEngineCore.so.6````
7. For Mac, install the package including ```libQt6WebEngineCore.so.6```, symlink the file to /usr/lib/python3.10/site-packages/PyQt6/Qt6/lib/libQt6WebEngineCore.so.6

**Reply by stan-ely** [reply]
Ubuntu 20.04 have an older version of pip in the [system repos](https://packages.ubuntu.com/focal/python3-pip) v20.0.2. This doesn't have PyQt6. I had to update pip as a user install.

```bash
pip install pip --update
```

**Reply by stan-ely** [reply]
Additional note. The Emacs in the repos version 26 if I remember is not supported again as json-parser is not available till Emacs 27. I used the Emacs available from snap.

**Comment by MatthewZMD** [comment]
EAF has officially upgraded its dependency to Qt6, please upgrade to Qt6 to use EAF's latest features.

However, we also understand that not everyone can upgrade to Qt6 due to various reasons, I've created a `qt5-legacy` branch that aims to keep a functional EAF under Qt5. Please `git checkout` this branch and run `install-eaf.py`.

I'll try to move future patches to this branch if it is backward compatible, but due to how incompatible Qt5 and Qt6 are, not every patch will be backward-compatible, and I cannot guarantee that this branch will be working forever.

Please upgrade to Qt6 and use the master branch for the latest features as soon as possible.

**Comment by MatthewZMD** [comment]
The EAF Browser has re-written how the cookies work, please therefore remove `~/.emacs.d/eaf/browser/cookie` directory after updating EAF.

**Comment by madorian** [comment]
The instructions for NixOS is missing Qt6. I get this on NixOs-22.05: 

```
Process *eaf* exited abnormally with code 1
Traceback (most recent call last):
  File "/home/b0ef/.emacs.d/site-lisp/emacs-application-framework/eaf.py", line 25, in <module>
    from PyQt6 import QtWebEngineWidgets as NeverUsed # noqa
ModuleNotFoundError: No module named 'PyQt6'

Process *eaf* exited abnormally with code 1
```

**Comment by manateelazycat** [comment]
The EAF Browser uses JavaScript's document.activeElement to detect whether a webpage is in input mode. When in input mode, it sends character events to the webpage. If it is in non-input mode (e.g., browsing Google), it enters single-key command mode (similar to vimium).

However, this detection method of the EAF Browser fails in the following scenarios:

1. Website security protection, such as when activeElement fails during QQ email login.
2. Cloud service backend, such as using an iframe to embed a terminal emulator in Google Cloud.
3. Various websites with overlays, such as during Reddit login, when a hidden floating window will snatch the activeElement.
4. 
To deal with these non-standard application scenarios, we have added a switch_to_input_mode command (Alt + Z) today, which forces the EAF Browser to enter input mode.

**Comment by manateelazycat** [comment]
Hi all, please upgrade to newest version by install-eaf.py

I have push new commit https://github.com/emacs-eaf/emacs-application-framework/commit/2eb66df5c1fc10103218746fca4a203e2d0055d1

This patch will avoid EAF crash when you search something in EAF application, such as, eaf-file-manager.
